crates/harp/src/size.rs

```diff
@@ -235,14 +238,61 @@ fn obj_size_tree(
                 return 0;
             }
 
-            size += obj_size_tree(
-                unsafe { libr::FRAME(x) },
-                base_env,
-                sizeof_node,
-                sizeof_vector,
-                seen,
-                depth + 1,
-            );
+            // We can't access environment bindings values using `CAR` because this requires knowledge
+            // about internals of environments.
+            // This also means we won't count the size of internal implementation details of environments,
+            // such as the pre-allocated size of hash tables, etc.
+
+            for binding in Environment::new(RObject::view(x))
+                .iter()
+                .filter_map(|x| x.ok())
+            {
+                // `binding.name`s are SYMSXP, for which we don't need to add to size.
+                // We do add a node size for each binding though, because that's
+                // the size the internal pairlist uses for each element.
+                size += sizeof_node;
+
+                size += match binding.value {
+                    // For active bindings, we compute the size of the function
+                    BindingValue::Active { fun } => {
+                        obj_size_tree(fun.sexp, base_env, sizeof_node, sizeof_vector, seen, depth)
+                    },
+                    // `obj_size_tree` is aware of altrep objects.
+                    BindingValue::Altrep { object, .. } => obj_size_tree(
+                        object.sexp,
+                        base_env,
+                        sizeof_node,
+                        sizeof_vector,
+                        seen,
+                        depth + 1,
+                    ),
+                    // `object_size_tree` is aware of promise objects.
+                    // The environment iterator will automatically return `PRVALUE` as
+                    // a `Standard` binding though. So this is only seeing unevaluated promises.
+                    // For evaluated promises, we are not counting the size of `PRCODE`, but hopefully
```

Comment:
No need to do now unless you want to, but the iterator should provide `PRCODE` too for promises. Luke agreed to provide promise accessors, as long as promises themselves are not exposed and instead accessed via the environment.